backend/src/ee/services/ssh/ssh-certificate-authority-fns.ts

```diff
@@ -0,0 +1,262 @@
+              return principal.endsWith(`.${baseDomain}`);
+            }
+            return principal === allowedHost;
+          })
+        ) {
+          throw new BadRequestError({
+            message: `Principal '${principal}' is not in the list of allowed hosts or domains.`
+          });
+        }
+      });
+      break;
+    }
+    default:
+      throw new BadRequestError({
+        message: "Failed to validate SSH certificate principals due to unrecognized requested certificate type"
+      });
+  }
+};
+
+/**
+ * Validate the requested SSH certificate TTL based on the SSH certificate template configuration.
+ * @param template - The SSH certificate template configuration
+ * @param ttl - The TTL to validate
+ * @returns The TTL (in seconds) to use for issuing the SSH certificate
+ */
+export const validateSshCertificateTtl = (template: TSshCertificateTemplates, ttl: string | undefined) => {
+  if (!ttl) {
+    // use default template ttl
+    return ms(template.ttl) / 1000;
+  }
+
+  if (ms(ttl) > ms(template.maxTTL)) {
+    throw new BadRequestError({
+      message: "Failed TTL validation due to TTL being greater than configured max TTL on template"
+    });
+  }
+
+  return ms(ttl) / 1000;
+};
+
+/**
+ * Create an SSH certificate for a user or host.
+ */
+export const createSshCert = ({ caPrivateKey, userPublicKey, keyId, principals, ttl, certType }: TCreateSshCertDTO) => {
+  const uniqueId = crypto.randomBytes(8).toString("hex");
+  const publicKeyFile = `user_key_${uniqueId}.pub`;
+  const privateKeyFile = `ssh_ca_key_${uniqueId}`;
+  const signedPublicKeyFile = `user_key_${uniqueId}-cert.pub`;
+
+  if (fs.existsSync(publicKeyFile)) fs.unlinkSync(publicKeyFile);
+  if (fs.existsSync(privateKeyFile)) fs.unlinkSync(privateKeyFile);
+  if (fs.existsSync(signedPublicKeyFile)) fs.unlinkSync(signedPublicKeyFile);
+
+  // write public and private keys to temp files
+  fs.writeFileSync(publicKeyFile, userPublicKey);
+  fs.writeFileSync(privateKeyFile, caPrivateKey);
+  fs.chmodSync(privateKeyFile, 0o600);
+
+  const serialNumber = createSshCertSerialNumber();
+
+  const certOptions = [
+    `-s ${privateKeyFile}`, // path to SSH CA private key
+    `-I "${keyId}"`, // identity for the issued certificate (key id)
+    `-n "${principals.join(",")}"`, // principal(s) that is user(s) or host(s)
+    `-V +${ttl}s`, // TTL in seconds (validity period) for the issue certificate
+    `-z ${serialNumber}`, // custom serial number for certificate
+    certType === "host" ? "-h" : "", // host certificate flag
+    publicKeyFile // path to signed [publicKey]
+  ]
+    .filter(Boolean)
+    .join(" ");
+
+  const command = `ssh-keygen ${certOptions}`;
+
+  // Execute the signing process
+  execSync(command);
+
+  const signedPublicKey = fs.readFileSync(signedPublicKeyFile, "utf8");
+
+  fs.unlinkSync(publicKeyFile);
+  fs.unlinkSync(privateKeyFile);
+  fs.unlinkSync(signedPublicKeyFile);
```

Comment:
same for this one here. we might be able to use `createCertificate` from sshpk?

<img width="889" alt="image" src="https://github.com/user-attachments/assets/2dca8101-f4f2-4f5f-9cc6-894b38935320">